BFD-3138: Add headers to OpenAPI spec yamls

rcrupp opened the pull request:
**JIRA Ticket:**
[BFD-3138](https://jira.cms.gov/browse/BFD-3138)

**User Story or Bug Summary:**
#### Background
The generated Open API specification for V1 and V2 via the HAPI FHIR Open API interceptor is a little lossy. It's missing some critical details such as custom BFD headers and other metadata that we want to use to enrich the outputs to make them more meaningful/useful.

---

### What Does This PR Do? 
* This code proposes a pattern for enhancing the OpenAPI yaml returned by the `/api-docs` endpoint.  
* The first enhancement is adding the BFD required and optional HTTP request headers.  
* The `OpenApiDocs` class has been modified to enrich the `/api-docs` yaml.
* A new yaml file, `headers.yaml`, was added to provide the header parameter specs.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:
* Verify all PR security questions and checklists have been completed and addressed.
* Verify that the all the necessary headers are included.  I only added `BlueButton-OriginalQueryId`, `BlueButton-DeveloperId`, `BlueButton-ApplicationId`,  `BlueButton-UserId`,  `IncludeIdentifiers`, `IncludeAddressFields`, and `IncludeTaxNumbers`, but if there are others listed on [Making-Requests-to-BFD](https://github.com/CMSgov/beneficiary-fhir-data/wiki/Making-Requests-to-BFD) I will add those.

* Verify that the `IncludeIdentifiers`, `IncludeAddressFields`, and `IncludeTaxNumbers` specs are included everywhere appropriate.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No

* If the answer to any of the questions below is **Yes**, then please add @<!-- -->StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A


### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] The data dictionary has been updated with any field mapping changes, if any were made.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

brianburton commented on apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/OpenApiDocs.java:
```diff
@@ -183,4 +190,70 @@ private void setRequestAuth() {
+    Map<String, Object> headers = yaml.load(headersYaml);
+
+    // add header params to be used in refs
+    var components = (Map<String, Object>) spec.get("components");
+    var headerComponents = (Map<String, Object>) headers.get("components");
+    var headerParameters = (Map<String, Object>) headerComponents.get("parameters");
```
I like the use of `var` here. Since a cast is needed anyway specifying the variable type explicitly would have just added redundancy.

brianburton reviewed (dismissed):
This looks great.  I like that the supported headers are config driven.

lsmitchell commented on apps/bfd-server/bfd-server-war/src/test/resources/openapi/headers.yaml:
```diff
@@ -0,0 +1,162 @@
+        type: string
+        format: date-time
+      description: | 
+        An ISO-8601 UTC timestamp representing (roughly) when the original request reached the frontend system.
```
These were probably originally written in the context of BB2 frontend, but I'm not sure if every partner is going to be a frontend system; some are intermediaries to an actual user-facing frontend system. I'm thinking we actually care when the request made it to the requestor's system, be it a frontend or intermediary.

Might need to ask a partner how they are populating this today, if they are actually taking an upstream timestamp from the original frontend requestor or just filling this out when their system gets the request. If the latter, we might want to update the verbiage on the description in these headers.

Despite the number of words here this isn't a huge deal, just a clarity thing, and could be handled in the future.

rcrupp replied:
Most of the timestamps I saw in splunk are within a couple hundred millisecs from timestamp on the log entry so it appears more like the latter.

lsmitchell commented on apps/bfd-server/bfd-server-war/src/test/resources/openapi/headers.yaml:
```diff
@@ -0,0 +1,162 @@
+      schema:
+        type: string
+      description: |
+        The unique beneficiary ID (e.g. Patient.id) from BFD for the user the beneficiary 
+        whose data is being requested.
```
If I were a consumer, I'd want to know why I'd need to use this header.. given that its not required and wed generally pass this in as a parameter if we were searching on bene id. (I don't actually know what this header is used for offhandedly)

mjburling replied:
🤔

rcrupp replied:
I can remove references to this one if it's not being used.

lsmitchell commented on apps/bfd-server/bfd-server-war/src/test/resources/openapi/headers.yaml:
```diff
@@ -0,0 +1,162 @@
+      required: false
+      schema:
+        type: string
+      description:  A unique identifier for the job (Job ID, UUID, etc.)
```
While the bulk headers aren't required, I'd assume we'd want to mention they are HIGHLY desired (or maybe conditionally required) on bulk jobs so we can properly track them in our system

mjburling replied:
They're either required or they aren't. Currently, they aren't. Even after this is published, I wouldn't expect anyone to read through all headers in the off-chance that the authors included a note urging that a header was _HIGHLY desired_. When this is sufficiently documented and socialized, we will begin planning for required inclusion of these headers, and update this documentation to be consistent with the implementation.

lsmitchell commented on apps/bfd-server/bfd-server-war/src/test/resources/openapi/headers.yaml:
```diff
@@ -0,0 +1,162 @@
+        default: "false"
+      description:  |
+        When set to true, BFD will include all of the detailed address data for the requested beneficiary. When set to 
+        false, BFD will not include detailed address data for the requested beneficiary. Please note that, even when 
+        false county and ZIP/postal codes will still be included for the specified beneficiary.
```
As far as I can see in the code (BeneficiaryTranformerV2:172) we actually return state and postalCode when this is false, not county.

rcrupp replied:
Corrected.

lsmitchell reviewed (commented):
Some small notes. Also worth double checking in Splunk (or the code) like we mentioned to ensure there arent any functional headers used that arent documented properly in the wiki.

rcrupp commented:
After checking the headers used in requests in Splunk I added the bulk headers which I had left out, made all the headers not required since that looks correct, and added those that I saw but hadn't included earlier.

lsmitchell commented on apps/utils/locust_tests/run-test.sh:
```diff
@@ -0,0 +1,30 @@
+export BFD_ENV="3085-test"
+export BFD_ROOT=/home/randall/code/beneficiary-fhir-data
```
Should this file be removed/renamed if this was a one-off test?

rcrupp replied:
Yes, removed
